plugins/woocommerce/includes/class-wc-order-item-product.php

```diff
@@ -177,8 +181,35 @@ public function set_taxes( $raw_tax_data ) {
 			'subtotal' => array(),
 		);
 		if ( ! empty( $raw_tax_data['total'] ) && ! empty( $raw_tax_data['subtotal'] ) ) {
-			$tax_data['subtotal'] = array_map( 'wc_format_decimal', $raw_tax_data['subtotal'] );
-			$tax_data['total']    = array_map( 'wc_format_decimal', $raw_tax_data['total'] );
+			$subtotal = $raw_tax_data['subtotal'];
+			$total    = $raw_tax_data['total'];
+
+			// Handle legacy data where total/subtotal might be floats/strings instead of arrays.
+			// Convert scalar values to array format to preserve the tax amount.
+			$has_legacy_data = ! is_array( $subtotal ) || ! is_array( $total );
+
+			if ( ! is_array( $subtotal ) ) {
+				$subtotal = $this->convert_legacy_tax_value_to_array( $subtotal );
+			}
+			if ( ! is_array( $total ) ) {
+				$total = $this->convert_legacy_tax_value_to_array( $total );
```

Comment:
nit: In cases where both total and subtotal are present, wouldn't this trigger a double fetch of the order? Maybe we can fetch the order here and pass as an arg?